src/analysis/seminorm.lean

```diff
@@ -45,51 +46,193 @@ open_locale big_operators nnreal pointwise topological_space
+
+variables {p q}
+
+@[to_additive] lemma comp_mono (hp : p ≤ q) : p.comp f ≤ q.comp f := λ _, hp _
+
+end group
+
+section comm_group
+variables [comm_group E] [comm_group F] (p q : group_seminorm E) (x y : E)
+
+/-- The direct path from `1` to `y` is shorter than the path with `x` "inserted" in between. -/
+@[to_additive "The direct path from `0` to `y` is shorter than the path with `x` \"inserted\" in
+between."]
+lemma le_insert : p y ≤ p x + p (x / y) :=
+calc p y = p (x / (x / y)) : by rw div_div_cancel
+     ... ≤ p x + p (x / y) : p.div_le  _ _
+
+/-- The direct path from 0 to x is shorter than the path with y "inserted" in between. -/
+@[to_additive "The direct path from 0 to x is shorter than the path with y \"inserted\" in
+between."]
+lemma le_insert' : p x ≤ p y + p (x / y) := by { rw div_rev, exact le_insert _ _ _ }
+
+@[to_additive] lemma comp_mul_le (f g : F →* E) : p.comp (f * g) ≤ p.comp f + p.comp g :=
+λ _, p.mul_le _ _
+
+@[to_additive] private lemma mul_bdd_below_range_add {p q : group_seminorm E} {x : E} :
+  bdd_below (range (λ y, p y + q (x / y))) :=
+⟨0, by { rintro _ ⟨x, rfl⟩, exact add_nonneg (p.nonneg _) (q.nonneg _) }⟩
+
+@[to_additive] noncomputable instance : has_inf (group_seminorm E) :=
+{ inf := λ p q,
+  { to_fun    := λ x, ⨅ y, p y + q (x / y),
+    nonneg'   := λ x, le_cinfi $ λ x, add_nonneg (p.nonneg _) (q.nonneg _),
+    map_one' := cinfi_eq_of_forall_ge_of_forall_gt_exists_lt
+        (λ x, add_nonneg (p.nonneg _) (q.nonneg _))
+        (λ r hr, ⟨1, by rwa [div_one, p.map_one, q.map_one, add_zero]⟩),
+    mul_le'   := λ x y, le_cinfi_add_cinfi $ λ u v, begin
+      apply cinfi_le_of_le mul_bdd_below_range_add (u * v),
+      dsimp,
+      rw [mul_div_mul_comm, add_add_add_comm],
+      exact add_le_add (p.mul_le _ _) (q.mul_le _ _),
+    end,
+    inv' := λ x, (inv_surjective.infi_comp _).symm.trans $ by simp_rw [p.inv, ←inv_div', q.inv] }}
+
+@[simp, to_additive] lemma inf_apply : (p ⊓ q) x = ⨅ y, p y + q (x / y) := rfl
+
+@[to_additive] noncomputable instance : lattice (group_seminorm E) :=
+{ inf := (⊓),
+  inf_le_left := λ p q x, cinfi_le_of_le mul_bdd_below_range_add x $
+    by rw [div_self', q.map_one, add_zero],
+  inf_le_right := λ p q x, cinfi_le_of_le mul_bdd_below_range_add (1 : E) $
+    by simp only [div_one, p.map_one, zero_add],
+  le_inf := λ a b c hb hc x, le_cinfi $ λ u, (a.le_insert' _ _).trans $ add_le_add (hb _) (hc _),
+  ..group_seminorm.semilattice_sup }
+
+end comm_group
+end group_seminorm
+
+namespace add_group_seminorm
+variables [add_group E] (p : add_group_seminorm E) (x y : E) (r : ℝ)
+
+instance zero_hom_class : zero_hom_class (add_group_seminorm E) E ℝ :=
+{ coe := λ f, f.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_zero := λ f, f.map_zero' }
 
 /-- Any action on `ℝ` which factors through `ℝ≥0` applies to an `add_group_seminorm`. -/
 instance [has_smul R ℝ] [has_smul R ℝ≥0] [is_scalar_tower R ℝ≥0 ℝ] :
```

Comment:
Why doesn't this generalize to the multiplicative case?

Reply:
It does, but `to_additive` is getting confused.